gio/src/cancellable_future.rs

```diff
@@ -87,49 +149,37 @@ where
     type Output = Result<<F as Future>::Output, Cancelled>;
 
     fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-        if self.is_cancelled() {
-            return Poll::Ready(Err(Cancelled));
+        if self.cancellable.is_cancelled() {
+            // XXX: Whenever we want to break the API, we should return here only
+            // self.cancellable.set_error_if_cancelled() value.
+            return std::task::Poll::Ready(Err(Cancelled));
         }
 
-        let mut this = self.as_mut().project();
+        let mut waker = self.waker.lock().unwrap();
+        if waker.is_none() {
+            *waker = Some(cx.waker().clone());
```

Comment:
Nit: `Arc::clone(&t)` is more explicit about why and how the global counter is used